[gold] :: Merging paid newsletters PRs.

n3f opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
This PR contains a series of fixes for the Paid Newsletter. We felt it best to combine the PRs below to test all the changes easily. A video of our walkthrough can be found at this link: p81Rsd-1fe-p2

Fixes #29859
Fixes #29710
Fixes #29862

Originally handled #29230 too, but waiting on fix/test.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. Sandbox `public-api.wordpress.com` and use the production store by disabling the test store in `0-sandbox.php`
2. Start the newsletter signup flow via launchpad by going to `[/setup/newsletter/intro](http://wordpress.com/setup/newsletter/intro)`
3. Once the site is created, **sandbox the site**
4. Create a post, and insert the subscribe block to confirm the “Add payments” button appears on the Subscribe block.
5. Confirm the newsletter access radio buttons are not visible in the Gutenberg panel
6. Go back to the launchpad and connect Stripe
8. Create a newsletter plan
9. Write another post
10. Set the access level to **paid-subscribers**
11. Change post visibility to password protected and verify that an error is shown in the newsletter access panel after entering a password
12. Change post visibility back to public
13. Publish the post for paid subscribers only
14. Open in an incognito window and verify that the user has no access
15. Subscribe using a real card and verify the content is unlocked
16. Subscribe to the same post again but instead select the free subscription option
17. Publish a post for all subscribers, and verify that all subscribers received the newsletter 
18. Publish another paid post, and verify that only the paid subscriber received the post

romarioraffington committed:
Create a different check for newsletter config

romarioraffington committed:
Gate components based on Newsletter config

romarioraffington committed:
Revert to running newsletter code when feature enabled

romarioraffington committed:
changelog

romarioraffington committed:
Add Jetpack_Memberships to is_newsletter_feature_enabled check

romarioraffington committed:
Fix fatal error

romarioraffington committed:
Merge branch 'trunk' into fix/display-add-payments-button-for-subscribe-block

millerf committed:
Update subscriptions.php

millerf committed:
Use 'everybody' instead of null

millerf committed:
Changelog

romarioraffington committed:
Display the correct text for locked newsletter post

romarioraffington committed:
changelog

romarioraffington committed:
Make the learn more link open in a new tab

n3f committed:
Merge remote-tracking branch 'origin/fix/newsletter-locked-text' into paid-news/nyc-merge

n3f committed:
Merge remote-tracking branch 'origin/fix/access-level' into paid-news/nyc-merge

romarioraffington commented on projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:
```diff
@@ -425,6 +432,11 @@ function render_block( $attributes ) {
 function get_post_access_level() {
+	if ( ! is_singular() ) {
+		// There is no "actual" current post.
+		return 'everybody';
```
```suggestion
		return Token_Subscription_Service::POST_ACCESS_LEVEL_EVERYBODY;
```

romarioraffington replied:
Added in [5d25c5e](https://github.com/Automattic/jetpack/pull/29860/commits/5d25c5e7225c50581a0412ce382b367d2e535fc7)

millerf reviewed (dismissed):
LGTM. Awesome work from the Gold team. Best team ever.

jeherve committed:
Fix comments

We used $$next_version$$ instead of $$next-version$$, which is not supported by the script.

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:
```diff
@@ -58,8 +59,14 @@ function register_block() {
 	}
 
-	if ( ! Jetpack_Gutenberg::is_newsletter_enabled() ) {
-		return; // Stop here if the Newsletter feature is not enabled.
+	/**
+	 * Do not proceed if the newsletter feature is not enabled
+	 */
+	if (
+		/** This filter is documented in class.jetpack-gutenberg.php */
+		! apply_filters( 'jetpack_subscriptions_newsletter_feature_enabled', false )
+	) {
```
Should we check if `class_exists( '\Jetpack_Memberships' )` here as well?

romarioraffington replied:
Added in [78bc0d0](https://github.com/Automattic/jetpack/pull/29860/commits/78bc0d02308cce7625df82d774f73a353ae0e5d6)!

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:
```diff
@@ -737,13 +751,21 @@ function maybe_gate_existing_comments( $comment ) {
  */
-function get_locked_content_placeholder_text() {
+function get_locked_content_placeholder_text( $newsletter_access_level ) {
+	$access_level = $newsletter_access_level === Token_Subscription_Service::POST_ACCESS_LEVEL_PAID_SUBSCRIBERS ? 'paid subscribers' : 'subscribers';
```
Should we make this 2 strings translatable?

romarioraffington replied:
Good catch! Fixed in [da399bf](https://github.com/Automattic/jetpack/pull/29860/commits/da399bfabdaf0f9c93db5092be80af83bf6a7e84)

jeherve commented on projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php:
```diff
@@ -737,13 +751,21 @@ function maybe_gate_existing_comments( $comment ) {
-			<h2 class="wp-block-heading has-text-align-center" style="margin-bottom:var(--wp--preset--spacing--60);font-size:24px">' . esc_html__( 'This post is for paid subscribers', 'jetpack' ) . '</h2>
+			<h2 class="wp-block-heading has-text-align-center" style="margin-bottom:var(--wp--preset--spacing--60);font-size:24px">' .
+
+			/* translators: Newsletter access level */
```
I think it would be nice to be a bit more verbose with that explanation, maybe by explicitly mentioning the 2 strings that can be added there?

romarioraffington replied:
Thanks for the feedback. Fixed in [292f836](https://github.com/Automattic/jetpack/pull/29860/commits/292f836e8f548114251aab76485cf32f9206c7b2)

romarioraffington committed:
Make transaction more verbose

romarioraffington committed:
Make string translatable

romarioraffington committed:
Add 'Jetpack_Memberships' check

romarioraffington committed:
Replace magic strings with const

n3f reviewed (commented):
LGTM, thanks @romarioraffington

jeherve commented:
@n3f It seems that this created a Fatal: #29882 
Could you take a look at this?

Thank you!

romarioraffington commented:
Noting here that the fatal was fixed in https://github.com/Automattic/jetpack/pull/29885
